kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionCoordinator.java

```diff
@@ -613,6 +627,10 @@ public void fail(Errors errors) {
                             if (epochAndMetadata.getCoordinatorEpoch() == coordinatorEpoch) {
                                     // This was attempted epoch fence that failed, so mark this state on the metadata
                                 epochAndMetadata.getTransactionMetadata().setHasFailedEpochFence(true);
+
+                                // this line is not present in Kafka code base ?
+                                epochAndMetadata.getTransactionMetadata().setPendingState(Optional.empty());
```

Comment:
This is the key point.

Reply:
Okay, but why do we need to set the pending state to empty?

Reply:
If don't reset the pending state, it seems that the transaction coordinator will throw a concurrent exception when re-end the transaction, refer to [here](https://github.com/streamnative/kop/blob/master/kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/transaction/TransactionCoordinator.java#L669), but there is no test to cover this case yet.